cli/src/ui.rs

```diff
@@ -87,7 +88,32 @@ impl BuiltinPager {
             pager,
             dynamic_pager_thread: std::thread::spawn(move || {
                 // This thread handles the actual paging.
-                minus::dynamic_paging(pager_handle).unwrap();
+                // TODO(#4182): handle error better
+                minus::dynamic_paging(pager_handle)
+                    .inspect_err(|err| {
+                        eprintln!("\n");
+                        if let minus::error::MinusError::Setup(
+                            minus::error::SetupError::InvalidTerminal,
+                        ) = err
+                        {
+                            eprintln!(indoc! {r#"
+                                jj ERROR: jj's builtin pager is incompatible with this terminal
+
+                                This is known to happen with `mintty`, the default Git Bash terminal on Windows.
+                                See also https://github.com/martinvonz/jj/issues/4182.
+
+                                POSSIBLE WORKAROUNDS:
+                                - Use `jj --no-pager`
+                                - Use a different terminal (e.g. Windows Terminal or the Command Prompt)
+                                - Configure a different pager, see https://martinvonz.github.io/jj/latest/windows/#pagination for Git Bash on Windows
+                                - Use `winpty jj ...` on Windows."#
+                                })
+                        } else {
+                            eprintln!("jj ERROR: built-in pager failed to start, will panic.")
+                        }
+                        eprintln!();
+                    })
```

Comment:
nit: Alternatively, you can send the error back through the join handle, and print it at `.finalize()` where `ui` is available.

https://github.com/yuja/jj/commit/dfb5251a6a996c3e99040c13eea547069c9216d4#diff-f72142370d81ec8e718184bed135927f6fb90a77c5f7f0195a99c40f48a4ee6eR106-R113
(This is an experimental patch to use streampager. It works, but brings some old crates as dependency.)

In any case, we don't have to `.unwrap()` the error.

Reply:
As I was complaining in https://github.com/martinvonz/jj/issues/4182#issuecomment-2259315065, the panic tends to happen early and abort any mutating operations. The only way I can think of easily preserving the error (which seems to match your suggestion), it will only be detected *after* the `jj` operation completes. So, jj could do some mutations, but all user output will be lost at that point. This seemed problematic to me, so while a panic is not great, it seems like a better alternative.

Reply:
As for switching to streampager, that would hopefully solve this problem as well.

However, I am not sure about it in its currently abandoned state. Despite the Discord discussion, I think somebody should run `cargo update` on it, and I think I remember some bugs when I was testing it out (though I don't remember whether I was just really picky; at that point I was wondering if I should use it instead of `less`).

Reply:
With https://github.com/martinvonz/jj/pull/4187, we could detect the error at the first `write` to the pager, but at that point, it's unclear to me how to best handle it and print a nice message to the user if the problem is with their terminal.

Reply:
> the panic tends to happen early and abort any mutating operations.

Well, the end result is basically the same behavior as `ui.pager="false"`. Panicking in the pager thread doesn't stop the main thread, and we'll need to fix #4187 anyway so the main thread can notice the problem early.

And, it's pager. Mutable operations usually happen before the pager is spawned.

Reply:
This contradicts a few of my assumptions/beliefs:

- I thought that the pager is initialized when the UI is, and I thought the UI is initialized in the very beginning of `jj` execution, before `cmd_whatever` is called.
- I thought a panic stops all threads. If not, we could replace it with something to stop all threads, perhaps?
- I thought that if `minus` fails, it becomes a black hole in which any data that has been written to it before or is written after is lost. Does it re-print its data store without paging? That would be nice.

Now, I'm not quite sure what's right or what's best; I think you are saying that at least the first two of my assumptions were wrong. If you can give me pointers to what's going on, I'd appreciate it, otherwise I'll look myself.

If we can get to a state where the behavior is the same as `ui.pager="false"`, with a warning, that would be ideal. Considering my last belief above, this seemed hard to achieve.

-------------

My main concern is that some sort of explanation is shown to people trying to use `jj` on Git Bash explaining to them that disabling the pager will cause things to start working.

Reply:
> * I thought that the pager is initialized when the UI is,

No. Pager starts as late as possible (so that early argument errors wouldn't be paged for example.)

> * I thought a panic stops all threads. If not, we could replace it with something to stop all threads, perhaps?

No. I'm not sure if we can inject panic to other running threads, but maybe we could abort the entire process if that's what we want.

> * I thought that if `minus` fails, it becomes a black hole in which any data that has been written to it before or is written after is lost. Does it re-print its data store without paging? That would be nice.

Yes and no. All data written to `ui` would be lost. New data can be printed after `ui.finalize_pager()`.

Reply:
Ok, I changed the whole thing to match the plan you suggested. I was in a mild rush, so there might be silly mistakes.